# 🚗 US Accidents Dataset — Data Cleaning & Feature Engineering
---
**Dataset:** US-Accidents: A Countrywide Traffic Accident Dataset  
**Author:** Sobhan Moosavi (Ohio State University)  
**Version:** V8 (2023) &nbsp;|&nbsp; **Records:** ~7.7 Million &nbsp;|&nbsp; **Period:** Feb 2016 – Mar 2023  
**License:** CC BY-NC-SA 4.0 (Research use only)

> **Notebook Goal:** Clean, validate, and feature-engineer the US Accidents dataset to make it analysis and ML-ready.

---


## 📦 1. Imports & Setup

In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')


In [2]:
df = pd.read_csv('/kaggle/input/us-accidents/US_Accidents.csv')

In [3]:
df2 = df.copy()

## 🔍 2. Dataset Overview
Getting an initial look at the data — shape, columns, sample rows, data types, missing values, and duplicates.


### 2.1 Head & Tail

In [3]:
df.head()

,ID,Source,Severity,Start_Time,End_Time,Start_Lat,Start_Lng,End_Lat,End_Lng,Distance(mi),...,Roundabout,Station,Stop,Traffic_Calming,Traffic_Signal,Turning_Loop,Sunrise_Sunset,Civil_Twilight,Nautical_Twilight,Astronomical_Twilight
0,A-1,Source2,3,2016-02-08 05:46:00,2016-02-08 11:00:00,39.865147,-84.058723,NaN,NaN,0.01,...,False,False,False,False,False,False,Night,Night,Night,Night
1,A-2,Source2,2,2016-02-08 06:07:59,2016-02-08 06:37:59,39.928059,-82.831184,NaN,NaN,0.01,...,False,False,False,False,False,False,Night,Night,Night,Day
2,A-3,Source2,2,2016-02-08 06:49:27,2016-02-08 07:19:27,39.063148,-84.032608,NaN,NaN,0.01,...,False,False,False,False,True,False,Night,Night,Day,Day
3,A-4,Source2,3,2016-02-08 07:23:34,2016-02-08 07:53:34,39.747753,-84.205582,NaN,NaN,0.01,...,False,False,False,False,False,False,Night,Day,Day,Day
4,A-5,Source2,2,2016-02-08 07:39:07,2016-02-08 08:09:07,39.627781,-84.188354,NaN,NaN,0.01,...,False,False,False,False,True,False,Day,Day,Day,Day


In [4]:
df.tail()

,ID,Source,Severity,Start_Time,End_Time,Start_Lat,Start_Lng,End_Lat,End_Lng,Distance(mi),...,Roundabout,Station,Stop,Traffic_Calming,Traffic_Signal,Turning_Loop,Sunrise_Sunset,Civil_Twilight,Nautical_Twilight,Astronomical_Twilight
7728389,A-7777757,Source1,2,2019-08-23 18:03:25,2019-08-23 18:32:01,34.00248,-117.37936,33.99888,-117.37094,0.543,...,False,False,False,False,False,False,Day,Day,Day,Day
7728390,A-7777758,Source1,2,2019-08-23 19:11:30,2019-08-23 19:38:23,32.76696,-117.14806,32.76555,-117.15363,0.338,...,False,False,False,False,False,False,Day,Day,Day,Day
7728391,A-7777759,Source1,2,2019-08-23 19:00:21,2019-08-23 19:28:49,33.77545,-117.84779,33.77740,-117.85727,0.561,...,False,False,False,False,False,False,Day,Day,Day,Day
7728392,A-7777760,Source1,2,2019-08-23 19:00:21,2019-08-23 19:29:42,33.99246,-118.40302,33.98311,-118.39565,0.772,...,False,False,False,False,False,False,Day,Day,Day,Day
7728393,A-7777761,Source1,2,2019-08-23 18:52:06,2019-08-23 19:21:31,34.13393,-117.23092,34.13736,-117.23934,0.537,...,False,False,False,False,False,False,Day,Day,Day,Day


### 2.2 Shape & Columns

In [5]:
df.shape

(7728394, 46)

In [6]:
df.columns

Index(['ID', 'Source', 'Severity', 'Start_Time', 'End_Time', 'Start_Lat',
       'Start_Lng', 'End_Lat', 'End_Lng', 'Distance(mi)', 'Description',
       'Street', 'City', 'County', 'State', 'Zipcode', 'Country', 'Timezone',
       'Airport_Code', 'Weather_Timestamp', 'Temperature(F)', 'Wind_Chill(F)',
       'Humidity(%)', 'Pressure(in)', 'Visibility(mi)', 'Wind_Direction',
       'Wind_Speed(mph)', 'Precipitation(in)', 'Weather_Condition', 'Amenity',
       'Bump', 'Crossing', 'Give_Way', 'Junction', 'No_Exit', 'Railway',
       'Roundabout', 'Station', 'Stop', 'Traffic_Calming', 'Traffic_Signal',
       'Turning_Loop', 'Sunrise_Sunset', 'Civil_Twilight', 'Nautical_Twilight',
       'Astronomical_Twilight'],
      dtype='str')

### 2.3 Random Sample

In [7]:
df.sample(10)

,ID,Source,Severity,Start_Time,End_Time,Start_Lat,Start_Lng,End_Lat,End_Lng,Distance(mi),...,Roundabout,Station,Stop,Traffic_Calming,Traffic_Signal,Turning_Loop,Sunrise_Sunset,Civil_Twilight,Nautical_Twilight,Astronomical_Twilight
6437065,A-6484588,Source1,2,2021-02-06 03:38:00,2021-02-06 05:17:50,43.092305,-75.231057,43.091925,-75.230209,0.050,...,False,False,False,False,False,False,Night,Night,Night,Night
3777376,A-3805658,Source1,2,2022-01-23 21:39:30,2022-01-23 22:05:00,45.055963,-93.286947,45.069637,-93.297294,1.071,...,False,False,False,False,False,False,Night,Night,Night,Night
1116136,A-1125911,Source2,2,2021-03-09 06:12:27,2021-03-09 09:13:22,26.543076,-80.188622,NaN,NaN,0.000,...,False,False,False,False,False,False,Night,Day,Day,Day
557775,A-561239,Source2,1,2022-07-20 08:13:31,2022-07-20 09:13:12,33.965092,-80.996155,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day
3523827,A-3533710,Source1,2,2016-10-26 17:10:04,2016-10-26 23:10:04,33.874110,-118.224740,33.874000,-118.229250,0.259,...,False,False,False,False,False,False,Day,Day,Day,Day
4528673,A-4562993,Source1,2,2022-07-06 17:54:00.000000000,2022-07-06 19:10:54.000000000,39.975862,-76.732606,39.975627,-76.733975,0.074,...,False,False,False,False,False,False,Day,Day,Day,Day
5361679,A-5402410,Source1,2,2022-08-23 16:01:00,2022-08-23 17:39:51,37.891363,-120.445546,37.890965,-120.445229,0.032,...,False,False,False,False,False,False,Day,Day,Day,Day
3277955,A-3287837,Source2,2,2017-09-05 16:51:51,2017-09-05 17:21:17,38.358131,-122.769623,NaN,NaN,0.000,...,False,False,False,False,True,False,Day,Day,Day,Day
4891933,A-4929022,Source1,2,2022-05-18 15:57:05.000000000,2022-05-18 18:04:08.000000000,35.733859,-78.650955,35.730511,-78.651186,0.232,...,False,True,False,False,False,False,Day,Day,Day,Day
5364732,A-5405485,Source1,2,2022-03-11 18:34:29.000000000,2022-03-11 19:54:10.000000000,32.835357,-117.197906,32.838665,-117.200195,0.264,...,False,False,False,False,False,False,Night,Night,Day,Day


### 2.4 Data Types & Info

In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7728394 entries, 0 to 7728393
Data columns (total 46 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   ID                     str    
 1   Source                 str    
 2   Severity               int64  
 3   Start_Time             str    
 4   End_Time               str    
 5   Start_Lat              float64
 6   Start_Lng              float64
 7   End_Lat                float64
 8   End_Lng                float64
 9   Distance(mi)           float64
 10  Description            str    
 11  Street                 str    
 12  City                   str    
 13  County                 str    
 14  State                  str    
 15  Zipcode                str    
 16  Country                str    
 17  Timezone               str    
 18  Airport_Code           str    
 19  Weather_Timestamp      str    
 20  Temperature(F)         float64
 21  Wind_Chill(F)          float64
 22  Humidity(%)            float6

### 2.5 Statistical Summary

In [9]:
df.describe()

,Severity,Start_Lat,Start_Lng,End_Lat,End_Lng,Distance(mi),Temperature(F),Wind_Chill(F),Humidity(%),Pressure(in),Visibility(mi),Wind_Speed(mph),Precipitation(in)
count,7.728394e+06,7.728394e+06,7.728394e+06,4.325632e+06,4.325632e+06,7.728394e+06,7.564541e+06,5.729375e+06,7.554250e+06,7.587715e+06,7.551296e+06,7.157161e+06,5.524808e+06
mean,2.212384e+00,3.620119e+01,-9.470255e+01,3.626183e+01,-9.572557e+01,5.618423e-01,6.166329e+01,5.825105e+01,6.483104e+01,2.953899e+01,9.090376e+00,7.685490e+00,8.407210e-03
std,4.875313e-01,5.076079e+00,1.739176e+01,5.272905e+00,1.810793e+01,1.776811e+00,1.901365e+01,2.238983e+01,2.282097e+01,1.006190e+00,2.688316e+00,5.424983e+00,1.102246e-01
min,1.000000e+00,2.455480e+01,-1.246238e+02,2.456601e+01,-1.245457e+02,0.000000e+00,-8.900000e+01,-8.900000e+01,1.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,2.000000e+00,3.339963e+01,-1.172194e+02,3.346207e+01,-1.177543e+02,0.000000e+00,4.900000e+01,4.300000e+01,4.800000e+01,2.937000e+01,1.000000e+01,4.600000e+00,0.000000e+00
50%,2.000000e+00,3.582397e+01,-8.776662e+01,3.618349e+01,-8.802789e+01,3.000000e-02,6.400000e+01,6.200000e+01,6.700000e+01,2.986000e+01,1.000000e+01,7.000000e+00,0.000000e+00
75%,2.000000e+00,4.008496e+01,-8.035368e+01,4.017892e+01,-8.024709e+01,4.640000e-01,7.600000e+01,7.500000e+01,8.400000e+01,3.003000e+01,1.000000e+01,1.040000e+01,0.000000e+00
max,4.000000e+00,4.900220e+01,-6.711317e+01,4.907500e+01,-6.710924e+01,4.417500e+02,2.070000e+02,2.070000e+02,1.000000e+02,5.863000e+01,1.400000e+02,1.087000e+03,3.647000e+01


### 2.6 Missing Values

In [10]:
df.isnull().sum()

ID                             0
Source                         0
Severity                       0
Start_Time                     0
End_Time                       0
Start_Lat                      0
Start_Lng                      0
End_Lat                  3402762
End_Lng                  3402762
Distance(mi)                   0
Description                    5
Street                     10869
City                         253
County                         0
State                          0
Zipcode                     1915
Country                        0
Timezone                    7808
Airport_Code               22635
Weather_Timestamp         120228
Temperature(F)            163853
Wind_Chill(F)            1999019
Humidity(%)               174144
Pressure(in)              140679
Visibility(mi)            177098
Wind_Direction            175206
Wind_Speed(mph)           571233
Precipitation(in)        2203586
Weather_Condition         173459
Amenity                        0
Bump      

### 2.7 Duplicate Rows

In [11]:
df.duplicated().sum()

np.int64(0)

## 📋 3. Dataset Summary & Column Descriptions

### 3.1 Dataset Overview

| Field | Details |
|---|---|
| **Official Name** | US-Accidents: A Countrywide Traffic Accident Dataset |
| **Author** | Sobhan Moosavi, Ohio State University |
| **Version** | V8 (latest as of 2023) |
| **Total Records** | ~7.7 million accidents |
| **Time Span** | February 2016 – March 2023 |
| **Geography** | 49 US States |
| **Format** | CSV |
| **License** | CC BY-NC-SA 4.0 (Research only) |


### 3.2 Column Descriptions (All 46 Columns)

| # | Column | Type | Description |
|---|---|---|---|
| 1 | `ID` | str | Unique alphanumeric record identifier (e.g., A-3456789). Primary key — no analytical value on its own. |
| 2 | `Source` | str | API/provider that reported the accident (MapQuest, Bing, or both). Useful for studying data collection bias. |
| 3 | `Severity` | int | Traffic impact score from 1 (minor delay) to 4 (road closure). Primary ML target — heavily skewed toward 2. |
| 4 | `Start_Time` | datetime | Exact timestamp when the accident was first detected. Rich source for time-based features. |
| 5 | `End_Time` | datetime | Timestamp when traffic flow returned to normal. Used to calculate accident duration. |
| 6 | `Start_Lat` | float | Latitude of the accident start point (24°–50° for contiguous US). Core geospatial column. |
| 7 | `Start_Lng` | float | Longitude of the accident start point (-125° to -65°). Negative values since US is in western hemisphere. |
| 8 | `End_Lat` | float | Latitude of the far end of the affected road segment. Often missing or equal to Start_Lat for minor accidents. |
| 9 | `End_Lng` | float | Longitude of the far end of the affected road segment. Works with End_Lat to define the impacted road segment. |
| 10 | `Distance(mi)` | float | Length of road (miles) affected by the accident. Proxy for severity — highways tend to have larger values. |
| 11 | `Description` | str | Auto-generated natural language summary from the API. Quality varies by source. Useful for NLP tasks. |
| 12 | `Street` | str | Street or highway name (e.g., I-95 N, Main St). Useful for identifying accident-prone roads. |
| 13 | `City` | str | City where the accident occurred. Miami, Houston, LA, Charlotte are the most frequent. |
| 14 | `County` | str | County of the accident. Useful for regional analysis between city and state granularity. |
| 15 | `State` | str | Two-letter US state abbreviation (e.g., CA, TX). One of the cleanest and most complete columns. |
| 16 | `Zipcode` | str | Postal ZIP code (5-digit or ZIP+4 format). Useful for hyper-local analysis. |
| 17 | `Country` | str | Always `US`. Zero variance — no analytical value. |
| 18 | `Timezone` | str | Timezone of accident location (US/Eastern, US/Central, US/Mountain, US/Pacific). Critical for time analysis. |
| 19 | `Airport_Code` | str | IATA code of nearest weather station (not nearest airport). Links each record to its weather readings. |
| 20 | `Weather_Timestamp` | datetime | Timestamp of the weather observation. May differ from Start_Time — check the gap before using. |
| 21 | `Temperature(F)` | float | Air temperature in Fahrenheit. Ranges ~-30°F to 120°F. Strong correlation with road conditions. |
| 22 | `Wind_Chill(F)` | float | Perceived temperature accounting for wind speed. Only meaningful below 50°F. High null rate in warm records. |
| 23 | `Humidity(%)` | float | Relative humidity (0–100%). High values often correlate with fog or rain. |
| 24 | `Pressure(in)` | float | Atmospheric pressure in inches of mercury (~29.92 inHg normal). Rapid drops indicate incoming storms. |
| 25 | `Visibility(mi)` | float | Horizontal visibility in miles (0 = whiteout, 10 = clear). Below 1 mile strongly linked to severe accidents. |
| 26 | `Wind_Direction` | str | Compass direction wind blows from (N, NE, E, … Calm, Variable). Abbreviated — needs expansion. |
| 27 | `Wind_Speed(mph)` | float | Wind speed in mph. Above 30 mph affects vehicle control. Contains outliers above 300 mph (data errors). |
| 28 | `Precipitation(in)` | float | Precipitation amount in inches. Nulls and zeros often mean the same thing (no precipitation). |
| 29 | `Weather_Condition` | str | Weather label at accident time (Clear, Light Rain, Heavy Snow, Fog, etc.). 100+ unique values — needs grouping. |
| 30 | `Amenity` | bool | POI flag: amenity (park, shop, restaurant) nearby. 76L False vs 1L True — highly imbalanced. |
| 31 | `Bump` | bool | POI flag: speed bump or hump nearby. Only ~3,500 True values out of 7.7M. |
| 32 | `Crossing` | bool | POI flag: pedestrian or road crossing nearby. ~8 lakh True values. |
| 33 | `Give_Way` | bool | POI flag: yield/give-way sign nearby. Only 36,582 True values — highly imbalanced. |
| 34 | `Junction` | bool | POI flag: road junction nearby. 571,342 True — one of the stronger road feature predictors. |
| 35 | `No_Exit` | bool | POI flag: no-exit/dead-end road nearby. Only 19,545 True values. |
| 36 | `Railway` | bool | POI flag: railway level crossing nearby. Only 66,979 True values. |
| 37 | `Roundabout` | bool | POI flag: roundabout nearby. Only 249 True values — most extreme imbalance in dataset. |
| 38 | `Station` | bool | POI flag: transit/bus station nearby. 201,901 True — high pedestrian and congestion zone indicator. |
| 39 | `Stop` | bool | POI flag: stop sign nearby. Useful for identifying stop-sign intersection accidents. |
| 40 | `Traffic_Calming` | bool | POI flag: traffic calming infrastructure nearby (chicanes, raised crosswalks, etc.). Only 7,598 True. |
| 41 | `Traffic_Signal` | bool | POI flag: traffic signal (traffic light) nearby. 1,143,772 True — most analytically important POI column. |
| 42 | `Turning_Loop` | bool | POI flag: turning loop nearby. All False in this dataset — zero variance, drop this column. |
| 43 | `Sunrise_Sunset` | str | Day/Night based on official sunrise/sunset times for that location and date. Clean and complete. |
| 44 | `Civil_Twilight` | str | Day/Night based on civil twilight (sun 0°–6° below horizon). Captures dawn/dusk transition. |
| 45 | `Nautical_Twilight` | str | Day/Night based on nautical twilight (sun 6°–12° below horizon). Darker than civil twilight. |
| 46 | `Astronomical_Twilight` | str | Day/Night based on astronomical twilight (sun 12°–18° below horizon). Closest to full night classification. |


### 3.3 Data Assessment

#### 🔴 Dirty Data Issues

| Category | Issue |
|---|---|
| `accuracy` | `Start_Time` and `End_Time` need dtype conversion to datetime |
| `completeness` | `End_Lat` has 3,402,762 null values |
| `completeness` | `End_Lng` has 3,402,762 null values |
| `accuracy` | `Distance(mi)` has 3,302,161 values = 0 |
| `accuracy` | Values in `Distance(mi)` are in miles — need to consider unit context |
| `completeness` | `Description` has 5 missing values |
| `completeness` | `Street` has 10,869 missing values |
| `completeness` | `City` has 253 missing values |
| `completeness` | `Zipcode` has 1,915 null values |
| `consistency` | `Zipcode` is object type — needs conversion to numeric |
| `completeness` | `Timezone` has 7,808 null values |
| `completeness` | `Airport_Code` has 22,635 null values |
| `accuracy` | `Weather_Timestamp` needs dtype conversion to datetime |
| `completeness` | `Weather_Timestamp` has 120,228 null values |
| `consistency` | `Temperature(F)` needs to be rounded to full integer |
| `completeness` | `Temperature(F)` has 163,853 null values |
| `completeness` | `Wind_Chill(F)` has 1,999,019 null values |
| `completeness` | `Humidity(%)` has 174,144 null values |
| `consistency` | `Humidity(%)` can be rounded and converted to int |
| `completeness` | `Pressure(in)` has 140,679 null values |
| `accuracy` | `Pressure(in)` can be rounded to int |
| `completeness` | `Visibility(mi)` has 177,098 null values |
| `consistency` | `Visibility(mi)` should be rounded to int |
| `consistency` | `Wind_Direction` uses abbreviations (e.g., E for East) — need full form |
| `completeness` | `Wind_Direction` has 175,206 null values |
| `completeness` | `Wind_Speed(mph)` has 571,233 null values |
| `accuracy` | `Wind_Speed(mph)` has 18 values above 300 mph — physically impossible |
| `accuracy` | 9 lakh rows have zero wind speed |
| `completeness` | `Precipitation(in)` has 2,203,586 missing values |
| `consistency` | `Weather_Condition` has 144 categories — 40 with count < 10 |
| `consistency` | 67 lakh rows across 9 main categories — remaining should be grouped as 'Other' |
| `completeness` | `Sunrise_Sunset` has 23,246 null values |
| `completeness` | `Civil_Twilight` has 23,246 null values |
| `completeness` | `Nautical_Twilight` has 23,246 null values |
| `completeness` | `Astronomical_Twilight` has 23,246 null values |

---

#### 🟡 Messy Data Issues

| Issue | Detail |
|---|---|
| `ID` format | Combination of `A-` prefix and row number — only numeric part is needed |
| `Description` content | Combines road name, timing, and short summary — varies by source |
| `Street`, `City`, `Country` | High cardinality — many unique values (informational) |
| `State` abbreviations | e.g., `CA` instead of `California` — needs full name mapping |
| `Zipcode` format | Either 5-digit or 9-digit with hyphen — needs standardization to 5-digit |
| `Timezone` format | Written as `US/Eastern` etc. — `US/` prefix needs to be stripped |
| `Wind_Direction` case | `Calm` and `CALM` are the same — inconsistent casing |
| `Accident Duration` | Can be derived from `Start_Time` and `End_Time` — not directly stored |
| All POI columns | Highly imbalanced True/False — Amenity (76L False), Roundabout (249 True), Turning_Loop (all False) |


## 🧪 4. Data Assessment (Column-by-Column Analysis)

Each column is assessed for:
- **Completeness** — missing/null values
- **Accuracy** — wrong or impossible values  
- **Consistency** — formatting, encoding, unit issues
- **Validity** — values within expected ranges

---


### 4.0 Assessment Trials & Cross-Column Checks

In [16]:
df.columns

Index(['ID', 'Source', 'Severity', 'Start_Time', 'End_Time', 'Start_Lat',
       'Start_Lng', 'End_Lat', 'End_Lng', 'Distance(mi)', 'Description',
       'Street', 'City', 'County', 'State', 'Zipcode', 'Country', 'Timezone',
       'Airport_Code', 'Weather_Timestamp', 'Temperature(F)', 'Wind_Chill(F)',
       'Humidity(%)', 'Pressure(in)', 'Visibility(mi)', 'Wind_Direction',
       'Wind_Speed(mph)', 'Precipitation(in)', 'Weather_Condition', 'Amenity',
       'Bump', 'Crossing', 'Give_Way', 'Junction', 'No_Exit', 'Railway',
       'Roundabout', 'Station', 'Stop', 'Traffic_Calming', 'Traffic_Signal',
       'Turning_Loop', 'Sunrise_Sunset', 'Civil_Twilight', 'Nautical_Twilight',
       'Astronomical_Twilight'],
      dtype='str')

In [11]:
df['Civil_Twilight'].isna().equals(df['Astronomical_Twilight'].isna())

True

In [10]:
df['Civil_Twilight'].isna().equals(df['Nautical_Twilight'].isna())

True

In [9]:
df['Civil_Twilight'].isna().equals(df['Sunrise_Sunset'].isna())

True

In [ ]:
df['Astronomical_Twilight'].isna().equals(df['Nautical_Twilight'].isna())

### 4.1 Astronomical Twilight

In [7]:
df['Astronomical_Twilight'].value_counts()

Astronomical_Twilight
Day      6377548
Night    1327600
Name: count, dtype: int64

In [4]:
df['Astronomical_Twilight'].info()

<class 'pandas.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Astronomical_Twilight
Non-Null Count    Dtype
--------------    -----
7705148 non-null  str  
dtypes: str(1)
memory usage: 84.5 MB


In [5]:
df['Astronomical_Twilight'].isna().sum()

np.int64(23246)

In [6]:
df['Astronomical_Twilight'].sample(10)

4150111      Day
6740637    Night
2070617      Day
1245670      Day
7365795      Day
1375698      Day
436188       Day
2115722      Day
4743919      Day
3519841      Day
Name: Astronomical_Twilight, dtype: str

### 4.2 Nautical Twilight

In [3]:
df['Nautical_Twilight'].value_counts()

Nautical_Twilight
Day      6076156
Night    1628992
Name: count, dtype: int64

In [117]:
df['Nautical_Twilight'].isna().sum()

np.int64(23246)

In [116]:
df['Nautical_Twilight'].sample(10)

5968316      Day
962728     Night
825083     Night
339643       Day
567743       Day
625568       Day
3357779      Day
2055315      Day
5239657      Day
5453511    Night
Name: Nautical_Twilight, dtype: str

### 4.3 Civil Twilight

In [115]:
df['Civil_Twilight'].isna().sum()

np.int64(23246)

In [114]:
df['Civil_Twilight'].value_counts()

Civil_Twilight
Day      5695619
Night    2009529
Name: count, dtype: int64

In [113]:
df['Civil_Twilight'].sample(10)

3239827      Day
1457015      Day
699283       Day
2096818      Day
4065937      Day
6778616    Night
2156707      Day
365354       Day
4815277      Day
805144       Day
Name: Civil_Twilight, dtype: str

### 4.4 Sunrise / Sunset

In [112]:
df['Sunrise_Sunset'].value_counts()

Sunrise_Sunset
Day      5334553
Night    2370595
Name: count, dtype: int64

In [111]:
df['Sunrise_Sunset'].isna().sum()

np.int64(23246)

In [109]:
df['Sunrise_Sunset'].sample(10)

3099154      Day
3875429      Day
1501084      Day
307025       Day
5778826      Day
7156806      Day
5441054    Night
5062225      Day
6948920      Day
4173562      Day
Name: Sunrise_Sunset, dtype: str

### 4.5 Turning Loop

In [108]:
df['Turning_Loop'].value_counts()

Turning_Loop
False    7728394
Name: count, dtype: int64

In [107]:
df['Turning_Loop'].isna().sum()

np.int64(0)

In [106]:
df['Turning_Loop'].sample(10)

6451998    False
1452111    False
6327585    False
3928515    False
2645959    False
477044     False
5159351    False
4151314    False
2853134    False
6956796    False
Name: Turning_Loop, dtype: bool

### 4.6 Traffic Signal

In [105]:
df['Traffic_Signal'].value_counts()

Traffic_Signal
False    6584622
True     1143772
Name: count, dtype: int64

In [104]:
df['Traffic_Signal'].isna().sum()

np.int64(0)

In [103]:
df['Traffic_Signal'].sample(10)

4789414    False
3832687    False
2948193    False
2828791    False
4916789    False
1764622    False
2854815     True
4117149    False
4651590    False
6436795    False
Name: Traffic_Signal, dtype: bool

### 4.7 Traffic Calming

In [98]:
df['Traffic_Calming'].isna().sum()

np.int64(0)

In [97]:
df['Traffic_Calming'].value_counts()

Traffic_Calming
False    7720796
True        7598
Name: count, dtype: int64

In [96]:
df['Traffic_Calming'].sample(10)

7216395    False
1470819    False
3518797    False
1505964    False
6643831    False
1079692    False
4331500    False
5679472    False
4897498    False
6757505    False
Name: Traffic_Calming, dtype: bool

### 4.8 Station

In [95]:
df['Station'].isna().sum()

np.int64(0)

In [94]:
df['Station'].value_counts()

Station
False    7526493
True      201901
Name: count, dtype: int64

In [93]:
df['Station'].sample(10)

6572623    False
6518536    False
6666151    False
496787     False
1604180    False
1492383    False
5663961    False
3515416    False
4014178    False
2640091    False
Name: Station, dtype: bool

### 4.9 Roundabout

In [92]:
df['Roundabout'].value_counts()

Roundabout
False    7728145
True         249
Name: count, dtype: int64

In [91]:
df['Roundabout'].info()

<class 'pandas.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Roundabout
Non-Null Count    Dtype
--------------    -----
7728394 non-null  bool 
dtypes: bool(1)
memory usage: 7.4 MB


In [90]:
df['Roundabout'].isna().sum()

np.int64(0)

In [89]:
df['Roundabout'].sample(10)

7539046    False
6883302    False
3789937    False
4470421    False
6595550    False
590245     False
1937545    False
4896699    False
1158701    False
3722317    False
Name: Roundabout, dtype: bool

### 4.10 Railway

In [88]:
df['Railway'].isna().sum()

np.int64(0)

In [87]:
df['Railway'].value_counts()

Railway
False    7661415
True       66979
Name: count, dtype: int64

In [86]:
df['Railway'].sample(10)

5379613    False
7070600    False
3508784    False
2421209    False
7679032    False
4325807    False
2185252    False
614356     False
7165665    False
5973644    False
Name: Railway, dtype: bool

### 4.11 No Exit

In [85]:
df['No_Exit'].describe()

count     7728394
unique          2
top         False
freq      7708849
Name: No_Exit, dtype: object

In [84]:
df['No_Exit'].info()

<class 'pandas.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: No_Exit
Non-Null Count    Dtype
--------------    -----
7728394 non-null  bool 
dtypes: bool(1)
memory usage: 7.4 MB


In [83]:
df['No_Exit'].value_counts()

No_Exit
False    7708849
True       19545
Name: count, dtype: int64

In [82]:
df['No_Exit'].sample(10)

2423291    False
820585     False
3445102    False
4196172    False
5244930    False
3755701    False
3746440    False
2287290    False
4759303    False
1060713    False
Name: No_Exit, dtype: bool

### 4.12 Junction

In [81]:
df['Junction'].isna().sum()

np.int64(0)

In [80]:
df['Junction'].value_counts()

Junction
False    7157052
True      571342
Name: count, dtype: int64

In [79]:
df['Junction'].sample(10)

5404404    False
7293747     True
5333412    False
16285       True
1712060     True
1952677    False
6827690    False
4747638    False
954716     False
7043516    False
Name: Junction, dtype: bool

### 4.13 Give Way

In [73]:
df['Give_Way'].value_counts()

Give_Way
False    7691812
True       36582
Name: count, dtype: int64

In [72]:
df['Give_Way'].sample(10)

2004056    False
7461111    False
2626282    False
1574296    False
5006467    False
4185288    False
6386859    False
3444293    False
6182650    False
7225568    False
Name: Give_Way, dtype: bool

### 4.14 Crossing

In [71]:
df['Crossing'].isna().sum()

np.int64(0)

In [69]:
df['Crossing'].value_counts()

Crossing
False    6854631
True      873763
Name: count, dtype: int64

In [68]:
df['Crossing'].sample(10)

1197579    False
2297254    False
6199740    False
6581609    False
2481218    False
3048455    False
6144296    False
5788522     True
5586462    False
6192278    False
Name: Crossing, dtype: bool

### 4.15 Bump

In [70]:
df['Bump'].isna().sum()

np.int64(0)

In [67]:
df['Bump'].value_counts()

Bump
False    7724880
True        3514
Name: count, dtype: int64

In [66]:
df['Bump'].sample(10)

6112895    False
4900204    False
1272526    False
5562947    False
3917515    False
4144445    False
5376804    False
5099385    False
404161     False
1133380    False
Name: Bump, dtype: bool

### 4.16 Amenity

In [64]:
df['Amenity'].info()

<class 'pandas.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Amenity
Non-Null Count    Dtype
--------------    -----
7728394 non-null  bool 
dtypes: bool(1)
memory usage: 7.4 MB


In [63]:
df['Amenity'].isna().sum()

np.int64(0)

In [62]:
df['Amenity'].value_counts()

Amenity
False    7632060
True       96334
Name: count, dtype: int64

In [61]:
df['Amenity'].sample(10)

4703156    False
400094     False
16632      False
4931470    False
5477122    False
7371986    False
3546487    False
5933696    False
2219280    False
5793541     True
Name: Amenity, dtype: bool

### 4.17 Weather Condition

In [50]:
temp_df.sort_values('count' , ascending=False).head(50)

,Weather_Condition,count
0,Fair,2560802
1,Mostly Cloudy,1016195
2,Cloudy,817082
3,Clear,808743
4,Partly Cloudy,698972
5,Overcast,382866
6,Light Rain,352957
7,Scattered Clouds,204829
8,Light Snow,128680
9,Fog,99238


In [31]:
temp_df = df['Weather_Condition'].value_counts().reset_index()

In [51]:
df['Weather_Condition'].value_counts()

Weather_Condition
Fair                         2560802
Mostly Cloudy                1016195
Cloudy                        817082
Clear                         808743
Partly Cloudy                 698972
                              ...   
Heavy Sleet / Windy                1
Sand / Windy                       1
Heavy Rain Shower / Windy          1
Blowing Snow Nearby                1
Drifting Snow                      1
Name: count, Length: 144, dtype: int64

In [17]:
df['Weather_Condition'].sample(10)

2014398             Fair
1255392           Cloudy
864919            Cloudy
539547              Fair
1065441           Cloudy
1808302             Fair
501345     Mostly Cloudy
230452              Rain
4596182             Fair
284360     Partly Cloudy
Name: Weather_Condition, dtype: str

### 4.18 Precipitation (in)

In [14]:
df['Precipitation(in)'].describe()

count    5.524808e+06
mean     8.407210e-03
std      1.102246e-01
min      0.000000e+00
25%      0.000000e+00
50%      0.000000e+00
75%      0.000000e+00
max      3.647000e+01
Name: Precipitation(in), dtype: float64

In [13]:
df['Precipitation(in)'].isna().sum()

np.int64(2203586)

In [11]:
df['Precipitation(in)'].sample(10)

2720870     NaN
2664479     NaN
303446      NaN
227524      NaN
5572850    0.02
2750352     NaN
5240946    0.00
1869721    0.00
3583672     NaN
6875932    0.00
Name: Precipitation(in), dtype: float64

### 4.19 Wind Speed (mph)

In [10]:
df[df['Wind_Speed(mph)'] == 0]

,ID,Source,Severity,Start_Time,End_Time,Start_Lat,Start_Lng,End_Lat,End_Lng,Distance(mi),...,Roundabout,Station,Stop,Traffic_Calming,Traffic_Signal,Turning_Loop,Sunrise_Sunset,Civil_Twilight,Nautical_Twilight,Astronomical_Twilight
43488,A-43489,Source2,3,2016-06-22 23:13:09,2016-06-22 23:58:09,34.000114,-117.063087,NaN,NaN,0.000,...,False,False,False,False,False,False,Night,Night,Night,Night
44266,A-44267,Source2,3,2016-06-25 05:47:46,2016-06-25 06:32:46,34.459503,-118.615616,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day
44410,A-44411,Source2,2,2016-06-26 05:41:28,2016-06-26 06:28:19,34.433743,-118.385757,NaN,NaN,0.000,...,False,False,False,False,False,False,Night,Day,Day,Day
44568,A-44569,Source2,2,2016-06-27 09:39:22,2016-06-27 10:18:00,34.365108,-118.503716,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day
44570,A-44571,Source2,2,2016-06-27 09:38:53,2016-06-27 10:15:00,34.402683,-118.454987,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7728358,A-7777726,Source1,2,2019-08-23 08:24:31,2019-08-23 08:52:30,34.120400,-117.740320,34.120190,-117.73288,0.426,...,False,False,False,False,False,False,Day,Day,Day,Day
7728361,A-7777729,Source1,2,2019-08-23 12:52:31,2019-08-23 13:20:14,34.023790,-118.276390,34.025760,-118.27529,0.150,...,False,False,False,False,False,False,Day,Day,Day,Day
7728364,A-7777732,Source1,2,2019-08-23 13:42:50,2019-08-23 14:10:06,34.070610,-118.263910,34.069740,-118.26155,0.148,...,False,False,False,False,False,False,Day,Day,Day,Day
7728375,A-7777743,Source1,2,2019-08-23 16:28:17,2019-08-23 16:57:04,34.323752,-118.500833,34.340840,-118.52113,1.654,...,False,False,False,False,False,False,Day,Day,Day,Day


In [16]:
df[df['Wind_Speed(mph)'] > 300]

,ID,Source,Severity,Start_Time,End_Time,Start_Lat,Start_Lng,End_Lat,End_Lng,Distance(mi),...,Roundabout,Station,Stop,Traffic_Calming,Traffic_Signal,Turning_Loop,Sunrise_Sunset,Civil_Twilight,Nautical_Twilight,Astronomical_Twilight
114580,A-114587,Source2,3,2016-06-13 18:28:43,2016-06-13 19:13:43,34.303253,-118.479774,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day
114588,A-114595,Source2,3,2016-06-13 18:44:57,2016-06-13 19:29:57,34.320553,-118.495262,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day
114633,A-114640,Source2,3,2016-06-13 20:06:45,2016-06-13 20:51:45,34.247803,-118.425835,NaN,NaN,0.000,...,False,False,False,False,False,False,Night,Day,Day,Day
114640,A-114647,Source2,2,2016-06-13 20:19:08,2016-06-13 21:04:08,34.266048,-118.467438,NaN,NaN,0.000,...,False,False,False,False,False,False,Night,Day,Day,Day
114647,A-114654,Source2,3,2016-06-13 20:30:47,2016-06-13 21:15:47,34.266048,-118.450768,NaN,NaN,0.000,...,False,False,False,False,False,False,Night,Day,Day,Day
2638811,A-2648683,Source2,2,2018-08-15 12:56:31,2018-08-15 13:26:16,38.559528,-121.373108,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day
2638812,A-2648684,Source2,2,2018-08-15 13:03:11,2018-08-15 13:32:56,38.557919,-121.386116,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day
3104760,A-3114639,Source2,2,2017-12-07 15:04:53,2017-12-07 15:49:38,34.030327,-117.650681,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day
3144938,A-3154817,Source2,3,2017-12-28 08:35:43,2017-12-28 09:05:27,34.279095,-118.469826,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day
3144960,A-3154839,Source2,3,2017-12-28 10:41:44,2017-12-28 11:26:26,34.321411,-118.496750,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day


In [15]:
df['Wind_Speed(mph)'].max()

np.float64(1087.0)

In [9]:
df['Wind_Speed(mph)'].describe()

count    7.157161e+06
mean     7.685490e+00
std      5.424983e+00
min      0.000000e+00
25%      4.600000e+00
50%      7.000000e+00
75%      1.040000e+01
max      1.087000e+03
Name: Wind_Speed(mph), dtype: float64

In [ ]:
df['Wind_Speed(mph)'].sample(5)

509928      4.6
741403     12.0
3200341     4.6
2412742     NaN
1065969     7.0
Name: Wind_Speed(mph), dtype: float64

In [8]:
df['Wind_Speed(mph)'].isnull().sum()

np.int64(571233)

### 4.20 Wind Direction

In [6]:
df['Wind_Direction'].info()

<class 'pandas.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Wind_Direction
Non-Null Count    Dtype
--------------    -----
7553188 non-null  str  
dtypes: str(1)
memory usage: 80.3 MB


In [5]:
df['Wind_Direction'].isnull().sum()

np.int64(175206)

In [4]:
df['Wind_Direction'].unique()

<ArrowStringArray>
[    'Calm',       'SW',      'SSW',      'WSW',      'WNW',       'NW',
     'West',      'NNW',      'NNE',    'South',    'North', 'Variable',
       'SE',      'SSE',      'ESE',     'East',       'NE',      'ENE',
        'E',        'W',        nan,        'S',      'VAR',     'CALM',
        'N']
Length: 25, dtype: str

In [3]:
df['Wind_Direction'].sample(5)

65487          Calm
2579100    Variable
5555249           W
6049078        CALM
7672956           S
Name: Wind_Direction, dtype: str

### 4.21 Visibility (mi)

In [114]:
df['Visibility(mi)'].isnull().sum()

np.int64(177098)

In [113]:
df['Visibility(mi)'].info()

<class 'pandas.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Visibility(mi)
Non-Null Count    Dtype  
--------------    -----  
7551296 non-null  float64
dtypes: float64(1)
memory usage: 59.0 MB


In [112]:
df['Visibility(mi)'].describe()

count    7.551296e+06
mean     9.090376e+00
std      2.688316e+00
min      0.000000e+00
25%      1.000000e+01
50%      1.000000e+01
75%      1.000000e+01
max      1.400000e+02
Name: Visibility(mi), dtype: float64

In [109]:
df['Visibility(mi)'].sample(10)

6482441    10.0
695495     10.0
764343     10.0
5733332     5.0
5064845    10.0
435873     10.0
6377248    10.0
6827935    10.0
6433039    10.0
6616943    10.0
Name: Visibility(mi), dtype: float64

### 4.22 Pressure (in)

In [108]:
df['Pressure(in)'].info()

<class 'pandas.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Pressure(in)
Non-Null Count    Dtype  
--------------    -----  
7587715 non-null  float64
dtypes: float64(1)
memory usage: 59.0 MB


In [107]:
df['Pressure(in)'].describe()

count    7.587715e+06
mean     2.953899e+01
std      1.006190e+00
min      0.000000e+00
25%      2.937000e+01
50%      2.986000e+01
75%      3.003000e+01
max      5.863000e+01
Name: Pressure(in), dtype: float64

In [106]:
df['Pressure(in)'].isnull().sum()

np.int64(140679)

In [103]:
df['Pressure(in)'].sample(10)

6494248    29.78
3725490    26.04
6221123    27.81
5151233    29.79
6974904    29.93
2115080    30.05
424181     29.56
4576206    29.74
3498555    30.04
3273386    29.99
Name: Pressure(in), dtype: float64

### 4.23 Humidity (%)

In [102]:
df['Humidity(%)'].info()

<class 'pandas.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Humidity(%)
Non-Null Count    Dtype  
--------------    -----  
7554250 non-null  float64
dtypes: float64(1)
memory usage: 59.0 MB


In [101]:
df['Humidity(%)'].describe()

count    7.554250e+06
mean     6.483104e+01
std      2.282097e+01
min      1.000000e+00
25%      4.800000e+01
50%      6.700000e+01
75%      8.400000e+01
max      1.000000e+02
Name: Humidity(%), dtype: float64

In [100]:
df['Humidity(%)'].isnull().sum()

np.int64(174144)

In [99]:
df['Humidity(%)'].sample(10)

2572720    82.0
2129075    84.0
5457182    46.0
1789052    75.0
1081598    24.0
3259979    69.0
796422     69.0
3163049    42.0
4159397    71.0
235317     92.0
Name: Humidity(%), dtype: float64

### 4.24 Wind Chill (F)

In [98]:
df['Wind_Chill(F)'].info()

<class 'pandas.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Wind_Chill(F)
Non-Null Count    Dtype  
--------------    -----  
5729375 non-null  float64
dtypes: float64(1)
memory usage: 59.0 MB


In [97]:
df['Wind_Chill(F)'].describe()

count    5.729375e+06
mean     5.825105e+01
std      2.238983e+01
min     -8.900000e+01
25%      4.300000e+01
50%      6.200000e+01
75%      7.500000e+01
max      2.070000e+02
Name: Wind_Chill(F), dtype: float64

In [96]:
df['Wind_Chill(F)'].isnull().sum()

np.int64(1999019)

In [95]:
df['Wind_Chill(F)'].sample(10)

867289     70.0
1597596    32.0
743637     17.0
5797028    83.0
6236232    85.0
6918555    64.0
5914832    96.0
219352      NaN
330293     18.6
1231867    45.0
Name: Wind_Chill(F), dtype: float64

### 4.25 Temperature (F)

In [94]:
df['Temperature(F)'].isnull().sum()

np.int64(163853)

In [93]:
df['Temperature(F)'].describe()

count    7.564541e+06
mean     6.166329e+01
std      1.901365e+01
min     -8.900000e+01
25%      4.900000e+01
50%      6.400000e+01
75%      7.600000e+01
max      2.070000e+02
Name: Temperature(F), dtype: float64

In [92]:
df['Temperature(F)'].sample(10)

7347928    65.0
1986478    75.0
7001623    88.0
5566496    77.0
7389721    57.9
1611806    42.0
628070     65.0
2628784    86.0
2731470    72.0
1152063    28.0
Name: Temperature(F), dtype: float64

### 4.26 Weather Timestamp

In [90]:
df['Weather_Timestamp'].describe()

count                 7608166
unique                 941331
top       2022-03-13 01:53:00
freq                     1311
Name: Weather_Timestamp, dtype: object

In [89]:
df['Weather_Timestamp'].isnull().sum()

np.int64(120228)

In [88]:
df['Weather_Timestamp'].sample(10)

2592632    2018-09-24 10:52:00
198092     2016-12-28 11:56:00
6801980    2020-12-07 19:15:00
5190368    2022-05-19 07:56:00
3887606    2022-04-08 00:38:00
3567577    2017-03-07 16:53:00
1011324    2021-06-02 19:53:00
5237671    2022-09-15 22:10:00
2882229    2018-04-18 07:53:00
718191     2022-01-31 19:53:00
Name: Weather_Timestamp, dtype: str

### 4.27 Airport Code

In [87]:
df['Airport_Code'].value_counts()

Airport_Code
KCQT    118332
KRDU    107267
KMCJ    101786
KBNA     98926
KCLT     97273
         ...  
K8D3         1
KVTN         1
KY22         1
KAVK         1
KICR         1
Name: count, Length: 2045, dtype: int64

In [86]:
df['Airport_Code'].isnull().sum()

np.int64(22635)

In [84]:
df['Airport_Code'].sample(10)

2346399    KJRB
1083207    KISM
2660590    KARW
6137171    KMGC
6691188    KLOM
887463     KBIV
3566758    KDMW
2023831    KE16
5395971    KPTV
1450676    KHHR
Name: Airport_Code, dtype: str

### 4.28 Timezone

In [83]:
df['Timezone'].sample(10)

7382844     US/Pacific
4414009     US/Pacific
552393      US/Eastern
3321030     US/Eastern
4152561     US/Pacific
4193752     US/Pacific
6884468     US/Eastern
7385511    US/Mountain
608599      US/Central
7727753    US/Mountain
Name: Timezone, dtype: str

In [82]:
df['Timezone'].value_counts()

Timezone
US/Eastern     3580167
US/Pacific     2062984
US/Central     1645616
US/Mountain     431819
Name: count, dtype: int64

In [81]:
df['Timezone'].isnull().sum()

np.int64(7808)

### 4.29 Zipcode

In [80]:
df['Zipcode'].value_counts()

Zipcode
91761         11247
91706         10022
92407          8922
92507          8850
33186          8375
              ...  
23430-7555        1
35077-5100        1
97827-9687        1
97022-8710        1
94401-2213        1
Name: count, Length: 825094, dtype: int64

In [79]:
df['Zipcode'].describe()

count     7726479
unique     825094
top         91761
freq        11247
Name: Zipcode, dtype: object

In [78]:
df['Zipcode'].isnull().sum()

np.int64(1915)

In [77]:
df['Zipcode'].sample(10)

6008530         08037
1435977    97212-5333
1379021    73112-5973
739543          23185
5058248         72113
1979239         23435
2014241    14619-1314
821261     23238-5009
1020625    98122-2319
5964224    33179-3113
Name: Zipcode, dtype: str

### 4.30 State

In [76]:
df['State'].info()

<class 'pandas.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: State
Non-Null Count    Dtype
--------------    -----
7728394 non-null  str  
dtypes: str(1)
memory usage: 73.7 MB


In [75]:
df['State'].value_counts()

State
CA    1741433
FL     880192
TX     582837
SC     382557
NY     347960
NC     338199
VA     303301
PA     296620
MN     192084
OR     179660
AZ     170609
GA     169234
IL     168958
TN     167388
MI     162191
LA     149701
NJ     140719
MD     140417
OH     118115
WA     108221
AL     101044
UT      97079
CO      90885
OK      83647
MO      77323
CT      71005
IN      67224
MA      61996
WI      34688
KY      32254
NE      28870
MT      28496
IA      26307
AR      22780
NV      21665
KS      20992
DC      18630
RI      16971
MS      15181
DE      14097
WV      13793
ID      11376
NM      10325
NH      10213
WY       3757
ND       3487
ME       2698
VT        926
SD        289
Name: count, dtype: int64

In [74]:
df['State'].isnull().sum()

np.int64(0)

In [73]:
df['State'].sample(10)

4940360    LA
7641793    WA
1104979    SC
931656     AL
2167157    NC
2147892    TX
4506033    FL
5421687    OR
139857     FL
2982156    CA
Name: State, dtype: str

### 4.31 County

In [72]:
df['County'].value_counts()

County
Los Angeles    526851
Miami-Dade     251601
Orange         241275
Harris         181196
Dallas         157024
                ...  
Mcminn              1
Glascock            1
Woods               1
Mellette            1
Ness                1
Name: count, Length: 1871, dtype: int64

In [71]:
df['County'].info()

<class 'pandas.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: County
Non-Null Count    Dtype
--------------    -----
7728394 non-null  str  
dtypes: str(1)
memory usage: 118.4 MB


In [70]:
df['County'].describe()

count         7728394
unique           1871
top       Los Angeles
freq           526851
Name: County, dtype: object

In [69]:
df['County'].isnull().sum()

np.int64(0)

In [68]:
df['County'].sample(10)

19536                Merced
1725244             Mathews
2658167      San Bernardino
3256267           San Diego
1780543    East Baton Rouge
2807734                Lake
7145782           Allegheny
3119388                Lake
4534382          Tuscaloosa
2826282          Providence
Name: County, dtype: str

### 4.32 City

In [67]:
df['City'].info()

<class 'pandas.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: City
Non-Null Count    Dtype
--------------    -----
7728141 non-null  str  
dtypes: str(1)
memory usage: 124.6 MB


In [66]:
df['City'].value_counts()

City
Miami                           186917
Houston                         169609
Los Angeles                     156491
Charlotte                       138652
Dallas                          130939
                                 ...  
Fairbank                             1
Pond Eddy                            1
Ness City                            1
Clarksdale                           1
American Fork-Pleasant Grove         1
Name: count, Length: 13678, dtype: int64

In [65]:
df['City'].isnull().sum()

np.int64(253)

In [64]:
df['City'].sample(10)

4614768         Phoenix
7418784        Pasadena
935070          Zebulon
5609805        Richmond
2565749          Malden
7209252         Slayton
4821846    Purcellville
4242558          Dayton
5253681         Ontario
7643331       Brentwood
Name: City, dtype: str

### 4.33 Street

In [62]:
df['Street'].value_counts().head(30)

Street
I-95 N      78430
I-95 S      74528
I-5 N       71968
I-10 E      55572
I-10 W      53725
I-5 S       51781
I-80 W      39662
I-80 E      36113
I-405 N     32364
I-75 N      28166
I-75 S      27546
I-90 E      26426
I-15 N      24470
I-94 W      24003
I-90 W      23279
US-101 N    20041
I-95        20028
I-94 E      19940
I-70 E      19697
I-15 S      19230
I-40 W      19034
I-84 E      18955
I-35 N      18761
I-405 S     18318
I-40 E      17800
CA-99 N     17651
I-70 W      17402
I-84 W      17311
I-64 E      16611
I-64 W      16072
Name: count, dtype: int64

In [60]:
df['Street'].describe()

count     7717525
unique     336306
top        I-95 N
freq        78430
Name: Street, dtype: object

In [59]:
df['Street'].isnull().sum()

np.int64(10869)

In [57]:
df['Street'].sample(10)

1811977    E Stockton Blvd
2034857        NC 50 Hwy N
276636           W 34th St
2943930         10th St NW
2014208             I-87 S
4353970             I-95 S
2994751       New Bern Ave
2261308            US-75 N
1154835        La Honda Rd
7162535            CA-24 W
Name: Street, dtype: str

### 4.34 Description

In [53]:
df['Description'].isnull().sum()

np.int64(5)

In [56]:
df['Description'].sample(10).values[0:5]

<ArrowStringArray>
[                                                       'Accident on SC-418 near Jenkins Bridge Rd.',
  'Stationary traffic on US-129 N from GA-82 (US-129) to GA-11 Bus/Lee St (US-129) due to accident.',
  'Right hand shoulder blocked due to accident on I-110 Northbound near I-405 Exits 36 37B 38A 38B.',
 'Entry ramp to Research Blvd Southbound from N Mopac Expy Southbound lane blocked due to accident.',
                                        'Incident on SC-324 near BRIDGEWATER RD Drive with caution.']
Length: 5, dtype: str

### 4.35 Distance (mi)

In [39]:
df['Distance(mi)'].describe()

count    7.728394e+06
mean     5.618423e-01
std      1.776811e+00
min      0.000000e+00
25%      0.000000e+00
50%      3.000000e-02
75%      4.640000e-01
max      4.417500e+02
Name: Distance(mi), dtype: float64

In [38]:
df['Distance(mi)'].isnull().sum()

np.int64(0)

In [37]:
df['Distance(mi)'][df['Distance(mi)'] == 0].count()

np.int64(3302161)

In [34]:
df['Distance(mi)'].sample(10)

6774279    0.541
323049     0.010
2594113    0.000
7084895    0.128
7599548    0.224
5452038    0.355
7142303    0.000
1069598    0.210
2822004    0.000
88329      0.000
Name: Distance(mi), dtype: float64

### 4.36 End Longitude

In [27]:
df['End_Lng'].describe()

count    4.325632e+06
mean    -9.572557e+01
std      1.810793e+01
min     -1.245457e+02
25%     -1.177543e+02
50%     -8.802789e+01
75%     -8.024709e+01
max     -6.710924e+01
Name: End_Lng, dtype: float64

In [26]:
df['End_Lng'].info()

<class 'pandas.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: End_Lng
Non-Null Count    Dtype  
--------------    -----  
4325632 non-null  float64
dtypes: float64(1)
memory usage: 59.0 MB


In [25]:
(df['End_Lng'].isna() == df['End_Lat'].isna()).all()

np.True_

In [17]:
demo1 = df['End_Lat'][df['End_Lat'].isna()].index

In [18]:
demo1

RangeIndex(start=0, stop=3402762, step=1)

In [19]:
demo2 = df['End_Lng'][df['End_Lng'].isna()].index

In [22]:
(demo1 == demo2)

array([ True,  True,  True, ...,  True,  True,  True], shape=(3402762,))

In [5]:
df['End_Lng'].isnull().sum()

np.int64(3402762)

In [4]:
df['End_Lng'].sample(5)

1981463          NaN
7335309   -87.609720
1671922          NaN
6875282   -95.365932
3855851   -76.726573
Name: End_Lng, dtype: float64

### 4.37 End Latitude

In [53]:
df['End_Lat'].max() , df['End_Lat'].min()

(np.float64(49.075), np.float64(24.566013))

In [52]:
df['End_Lat'].describe()

count    4.325632e+06
mean     3.626183e+01
std      5.272905e+00
min      2.456601e+01
25%      3.346207e+01
50%      3.618349e+01
75%      4.017892e+01
max      4.907500e+01
Name: End_Lat, dtype: float64

In [51]:
df['End_Lat'].info()

<class 'pandas.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: End_Lat
Non-Null Count    Dtype  
--------------    -----  
4325632 non-null  float64
dtypes: float64(1)
memory usage: 59.0 MB


In [ ]:
df['End_Lat'].isnull().sum()

np.int64(3402762)

In [49]:
df['End_Lat'].sample(5)

1618093         NaN
3287254         NaN
1506747         NaN
3832752    32.79233
2282924         NaN
Name: End_Lat, dtype: float64

### 4.38 Start Longitude

In [41]:
df['Start_Lng'].max() , df['Start_Lng'].min()

(np.float64(-67.113167), np.float64(-124.623833))

In [40]:
df['Start_Lng'].describe()

count    7.728394e+06
mean    -9.470255e+01
std      1.739176e+01
min     -1.246238e+02
25%     -1.172194e+02
50%     -8.776662e+01
75%     -8.035368e+01
max     -6.711317e+01
Name: Start_Lng, dtype: float64

In [39]:
df['Start_Lng'].isnull().sum()

np.int64(0)

In [37]:
df['Start_Lng'].info()

<class 'pandas.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Start_Lng
Non-Null Count    Dtype  
--------------    -----  
7728394 non-null  float64
dtypes: float64(1)
memory usage: 59.0 MB


In [36]:
df['Start_Lng'].sample(5)

4368599   -80.239452
6377458   -85.270997
7023510   -87.947060
2925339   -84.398369
6116676   -77.497183
Name: Start_Lng, dtype: float64

### 4.39 Start Latitude

In [28]:
df['Start_Lat'].isnull().sum()

np.int64(0)

In [27]:
df['Start_Lat'].max() , df['Start_Lat'].min()

(np.float64(49.002201), np.float64(24.5548))

In [26]:
df['Start_Lat'].info()

<class 'pandas.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Start_Lat
Non-Null Count    Dtype  
--------------    -----  
7728394 non-null  float64
dtypes: float64(1)
memory usage: 59.0 MB


In [25]:
df['Start_Lat'].sample(5)

4929941    34.014335
6246358    48.543386
3859447    40.631782
3598924    32.771821
1206398    34.716228
Name: Start_Lat, dtype: float64

### 4.40 End Time

In [22]:
df['End_Time'].isnull().sum()

np.int64(0)

In [21]:
df['End_Time'].info()

<class 'pandas.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: End_Time
Non-Null Count    Dtype
--------------    -----
7728394 non-null  str  
dtypes: str(1)
memory usage: 59.0 MB


In [20]:
df['End_Time'].sample(5)

6705406              2020-09-28 16:18:30
3477102              2016-08-05 20:53:07
4213671    2022-01-31 16:00:41.000000000
1121655              2021-03-04 18:12:42
2216858              2019-03-19 19:43:54
Name: End_Time, dtype: str

### 4.41 Start Time

In [17]:
df['Start_Time'].isnull().sum()

np.int64(0)

In [16]:
df['Start_Time'].describe()

count                 7728394
unique                6131796
top       2021-01-26 16:16:13
freq                      225
Name: Start_Time, dtype: object

In [15]:
df['Start_Time'].info()

<class 'pandas.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Start_Time
Non-Null Count    Dtype
--------------    -----
7728394 non-null  str  
dtypes: str(1)
memory usage: 59.0 MB


In [14]:
df['Start_Time'].sample(5)

3512013    2016-06-17 17:32:05
5128878    2022-10-05 17:14:00
2765162    2018-06-21 13:10:50
3393506    2017-08-25 17:03:04
6229098    2021-11-02 17:34:01
Name: Start_Time, dtype: str

### 4.42 Severity

In [35]:
df['Severity'].isnull().sum()

np.int64(0)

In [36]:
df['Severity'].info()

<class 'pandas.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Severity
Non-Null Count    Dtype
--------------    -----
7728394 non-null  int64
dtypes: int64(1)
memory usage: 59.0 MB


In [37]:
df['Severity'].describe()

count    7.728394e+06
mean     2.212384e+00
std      4.875313e-01
min      1.000000e+00
25%      2.000000e+00
50%      2.000000e+00
75%      2.000000e+00
max      4.000000e+00
Name: Severity, dtype: float64

In [38]:
df['Severity'].value_counts()

Severity
2    6156981
3    1299337
4     204710
1      67366
Name: count, dtype: int64

In [39]:
df['Severity'].duplicated().sum()

np.int64(7728390)

In [40]:
df['Severity'].head()

0    3
1    2
2    2
3    3
4    2
Name: Severity, dtype: int64

In [42]:
df['Severity'].tail()

7728389    2
7728390    2
7728391    2
7728392    2
7728393    2
Name: Severity, dtype: int64

In [54]:
df['Severity'].sample(10)

5285235    2
248391     2
1324526    3
4093722    2
4896069    2
5463318    2
5836668    2
1651555    3
599773     3
1358504    2
Name: Severity, dtype: int64

### 4.43 Source

In [17]:
df['Source'].isnull().sum()

np.int64(0)

In [18]:
df['Source'].info()

<class 'pandas.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Source
Non-Null Count    Dtype
--------------    -----
7728394 non-null  str  
dtypes: str(1)
memory usage: 59.0 MB


In [19]:
df['Source'].value_counts()

Source
Source1    4325632
Source2    3305373
Source3      97389
Name: count, dtype: int64

In [20]:
df['Source'].describe()

count     7728394
unique          3
top       Source1
freq      4325632
Name: Source, dtype: object

In [21]:
df['Source'].head()

0    Source2
1    Source2
2    Source2
3    Source2
4    Source2
Name: Source, dtype: str

In [22]:
df['Source'].tail()

7728389    Source1
7728390    Source1
7728391    Source1
7728392    Source1
7728393    Source1
Name: Source, dtype: str

In [34]:
df['Source'].sample(10)

4951563    Source1
4992932    Source1
4359823    Source1
7047089    Source1
7640391    Source1
5293641    Source1
2114947    Source2
6157832    Source1
1771431    Source2
3886687    Source1
Name: Source, dtype: str

---
## 🧹 5. Data Cleaning

> Working on `df2` (a copy of the original `df`) to preserve raw data.

### Issues Identified & Actions Taken

| Category | Issues Found | Action |
|---|---|---|
| **Accuracy** | Wrong dtypes for datetime columns | Converted to `datetime64` |
| **Completeness** | Nulls in End_Lat, End_Lng (34L rows) | Filled with Start_Lat/Lng |
| **Completeness** | Nulls in weather columns | Filled with median/mode |
| **Consistency** | Wind direction abbreviations | Expanded to full names |
| **Consistency** | State abbreviations | Mapped to full state names |
| **Consistency** | Timezone format `US/Eastern` | Stripped to `Eastern` |
| **Accuracy** | Wind speed > 300 mph (18 rows) | Clipped to 300 |
| **Validity** | Zipcode as object type | Converted to int32 |

---


### 5.1 Handling Completeness — Filling Null Values

#### End_Lat & End_Lng
> **Strategy:** Filled using `Start_Lat` and `Start_Lng`.  
> **Justification:** 29 out of 34 lakh missing rows have `Distance = 0`, meaning start and end point are the same. For remaining rows, no bearing data is available so start coordinates are the best proxy.


In [4]:
# Verify this overlap
missing_end = df2[df2['End_Lat'].isna()]
print(missing_end['Distance(mi)'].value_counts().head())
# Most likely Distance = 0 dominates here

Distance(mi)
0.00    2900689
0.01     249056
0.01      13367
0.02       6011
0.45       1144
Name: count, dtype: int64


In [5]:
df2['End_Lat'].isna().sum()

np.int64(3402762)

In [6]:
df2['End_Lat'] = df2['End_Lat'].fillna(df2['Start_Lat'])

In [7]:
df2['End_Lat'].isnull().sum()

np.int64(0)

In [8]:
df2['End_Lng'] = df2['End_Lng'].fillna(df2['Start_Lng'])

In [9]:
df2['End_Lng'].isnull().sum()

np.int64(0)

#### Description
> **Strategy:** Filled with `'No Information'` — 5 missing values only.


In [10]:
df2['Description'].isna().sum()

np.int64(5)

In [11]:
df2['Description'] = df2['Description'].fillna('No Information')

In [12]:
df2['Description'].isna().sum()

np.int64(0)

#### Street
> **Strategy:** Filled with `'Not Known'` — 10,869 missing values.


In [13]:
df2['Street'].isna().sum()

np.int64(10869)

In [14]:
df2['Street'] = df2['Street'].fillna('Not Known')

In [15]:
df2['Street'].isna().sum()

np.int64(0)

#### City
> **Strategy:** Filled with `'Not Known'` — 253 missing values.


In [16]:
df2['City'].isna().sum()

np.int64(253)

In [17]:
df2['City'] = df2['City'].fillna('Not Known')

In [18]:
df2['City'].isna().sum()

np.int64(0)

#### Zipcode
> **Strategy:** Filled with `'0'` before type conversion — 1,915 null values.


In [19]:
df2['Zipcode']= df2['Zipcode'].fillna('0')

In [20]:
df2['Zipcode'].isnull().sum()

np.int64(0)

#### Timezone
> **Strategy:** Derived from longitude ranges since the US has only 4 major timezones.  
> `lng >= -80` → Eastern, `lng >= -96` → Central, `lng >= -115` → Mountain, else → Pacific


In [21]:
df2['Timezone'].isnull().sum()

np.int64(7808)

In [22]:
## works on small dataset


# def fillTimezone(row):

#     if pd.isna(row['Timezone']):
#         lng = row['Start_Lng']

#         if lng >= -80: return 'US/Eastern'
#         elif lng >= -96: return 'US/Central'
#         elif lng >= -115: return 'US/Mountain'
#         else : return 'US/Pacific'

#     return row['Timezone']

In [23]:
conditions = [
    df2['Start_Lng'] >= -80,
    df2['Start_Lng'] >= -96,
    df2['Start_Lng'] >= -115,
    df2['Start_Lng'] < -115
]

choices = ['US/Eastern', 'US/Central', 'US/Mountain', 'US/Pacific']

df2['Timezone'] = df2['Timezone'].fillna(
    pd.Series(np.select(conditions, choices, default='US/Pacific'), index=df2.index)
)

In [24]:
df2['Timezone'].isnull().sum()

np.int64(0)

#### Airport Code
> **Strategy:** Filled with `'Not Known'` — 22,635 null values.


In [25]:
df2['Airport_Code'].isna().sum()

np.int64(22635)

In [26]:
df2['Airport_Code'] = df2['Airport_Code'].fillna('Not Known')

In [27]:
df2['Airport_Code'].isna().sum()

np.int64(0)

#### Weather Timestamp
> **Strategy:** Filled with `Start_Time` — 120,228 null values. Weather at accident time is the best approximation.


In [28]:
df2['Weather_Timestamp'].isnull().sum()

np.int64(120228)

In [29]:
df2['Weather_Timestamp'] = df2['Weather_Timestamp'].astype('datetime64[ns]')

In [30]:
df2['Weather_Timestamp'] = df2['Weather_Timestamp'].fillna(df2['Start_Time'])

In [31]:
df2['Weather_Timestamp'].isnull().sum()

np.int64(0)

#### Numeric Weather Columns
> **Strategy:** Filled with **median** (not mean) because these columns have outliers and skewed distributions.  
> Columns: `Temperature(F)`, `Wind_Chill(F)`, `Humidity(%)`, `Pressure(in)`, `Visibility(mi)`, `Wind_Speed(mph)`, `Precipitation(in)`


In [32]:
df2['Temperature(F)'].isna().sum()

np.int64(163853)

In [33]:
df2['Temperature(F)'] = df2['Temperature(F)'].fillna(df2['Temperature(F)'].median())

In [34]:
df2['Temperature(F)'].isna().sum()

np.int64(0)

#### Wind_Chill

In [35]:
df2['Wind_Chill(F)'].isna().sum()

np.int64(1999019)

In [36]:
df2['Wind_Chill(F)'] = df2['Wind_Chill(F)'].fillna(df2['Wind_Chill(F)'].median() )

In [37]:
df2['Wind_Chill(F)'].isna().sum()

np.int64(0)

#### 'Humidity(%)' , 'Pressure(in)', 'Visibility(mi)' , 'Wind_Speed(mph)', 'Precipitation(in)' , 'Weather_Condition' , 'Wind_Direction'

In [38]:
Num_cols = ['Humidity(%)' , 'Pressure(in)', 'Visibility(mi)' , 'Wind_Speed(mph)', 'Precipitation(in)' ]
cat_cols = ['Weather_Condition' , 'Wind_Direction']

In [39]:
for col in Num_cols:
    print(col , ':' , df2[col].isnull().sum())

Humidity(%) : 174144
Pressure(in) : 140679
Visibility(mi) : 177098
Wind_Speed(mph) : 571233
Precipitation(in) : 2203586


In [40]:
for col in Num_cols:
    df2[col] = df2[col].fillna(df2[col].median())

In [41]:
for col in Num_cols:
    print(col , ':' , df2[col].isnull().sum())

Humidity(%) : 0
Pressure(in) : 0
Visibility(mi) : 0
Wind_Speed(mph) : 0
Precipitation(in) : 0


In [42]:
for col in cat_cols:
    print(col , ':' , df2[col].isnull().sum())

Weather_Condition : 173459
Wind_Direction : 175206


In [43]:
for col in cat_cols:
    df2[col] = df2[col].fillna(df2[col].mode()[0])

In [44]:
for col in cat_cols:
    print(col , ':' , df2[col].isnull().sum())

Weather_Condition : 0
Wind_Direction : 0


#### 'Sunrise_Sunset', 'Civil_Twilight', 'Nautical_Twilight', 'Astronomical_Twilight'

In [45]:
cols = ['Sunrise_Sunset', 'Civil_Twilight', 'Nautical_Twilight',
       'Astronomical_Twilight']

In [46]:
for col in cols:
    print( col ,':' , df2[col].isna().sum())

Sunrise_Sunset : 23246
Civil_Twilight : 23246
Nautical_Twilight : 23246
Astronomical_Twilight : 23246


In [47]:
for col in cols:
    df2[col] = df2[col].fillna(df2[col].mode()[0])

In [48]:
for col in cols:
    print( col ,':' , df2[col].isna().sum())

Sunrise_Sunset : 0
Civil_Twilight : 0
Nautical_Twilight : 0
Astronomical_Twilight : 0


#### Twilight & Sunrise/Sunset Columns  
> **Strategy:** Filled with **mode** — categorical Day/Night columns, mode is appropriate.


---
### 5.2 Handling Messy Data — Consistency & Formatting Fixes


#### ID Column
> Stripped the `A-` prefix, keeping only the numeric part. The letter prefix adds no analytical value.


In [49]:
df2['ID']

0                A-1
1                A-2
2                A-3
3                A-4
4                A-5
             ...    
7728389    A-7777757
7728390    A-7777758
7728391    A-7777759
7728392    A-7777760
7728393    A-7777761
Name: ID, Length: 7728394, dtype: str

In [50]:
df2['ID'] = df2['ID'].str[2:]


In [51]:
df2['ID']

0                1
1                2
2                3
3                4
4                5
            ...   
7728389    7777757
7728390    7777758
7728391    7777759
7728392    7777760
7728393    7777761
Name: ID, Length: 7728394, dtype: str

#### State Column
> Mapped abbreviations to full state names. e.g., `CA` → `California`, `TX` → `Texas`.


In [52]:
df2['State']

0          OH
1          OH
2          OH
3          OH
4          OH
           ..
7728389    CA
7728390    CA
7728391    CA
7728392    CA
7728393    CA
Name: State, Length: 7728394, dtype: str

In [53]:
us_states = {
    'AL': 'Alabama', 'AK': 'Alaska', 'AZ': 'Arizona', 'AR': 'Arkansas',
    'CA': 'California', 'CO': 'Colorado', 'CT': 'Connecticut', 'DE': 'Delaware',
    'FL': 'Florida', 'GA': 'Georgia', 'HI': 'Hawaii', 'ID': 'Idaho',
    'IL': 'Illinois', 'IN': 'Indiana', 'IA': 'Iowa', 'KS': 'Kansas',
    'KY': 'Kentucky', 'LA': 'Louisiana', 'ME': 'Maine', 'MD': 'Maryland',
    'MA': 'Massachusetts', 'MI': 'Michigan', 'MN': 'Minnesota', 'MS': 'Mississippi',
    'MO': 'Missouri', 'MT': 'Montana', 'NE': 'Nebraska', 'NV': 'Nevada',
    'NH': 'New Hampshire', 'NJ': 'New Jersey', 'NM': 'New Mexico', 'NY': 'New York',
    'NC': 'North Carolina', 'ND': 'North Dakota', 'OH': 'Ohio', 'OK': 'Oklahoma',
    'OR': 'Oregon', 'PA': 'Pennsylvania', 'RI': 'Rhode Island', 'SC': 'South Carolina',
    'SD': 'South Dakota', 'TN': 'Tennessee', 'TX': 'Texas', 'UT': 'Utah',
    'VT': 'Vermont', 'VA': 'Virginia', 'WA': 'Washington', 'WV': 'West Virginia',
    'WI': 'Wisconsin', 'WY': 'Wyoming', 'DC': 'District of Columbia'
}

In [54]:
df2['State'] = df2['State'].map(us_states)

#### Zipcode Column
> Standardized to 5-digit format (trimmed 9-digit hyphenated codes) and converted to `int32`.


In [55]:
df2['Zipcode']

0               45424
1          43068-3402
2               45176
3               45417
4               45459
              ...    
7728389         92501
7728390         92108
7728391         92866
7728392         90230
7728393         92346
Name: Zipcode, Length: 7728394, dtype: str

In [56]:
df2['Zipcode'] = df2['Zipcode'].str[0:5]

In [57]:
df2['Zipcode'] = df2['Zipcode'].astype('int32')

In [58]:
df2['Zipcode']

0          45424
1          43068
2          45176
3          45417
4          45459
           ...  
7728389    92501
7728390    92108
7728391    92866
7728392    90230
7728393    92346
Name: Zipcode, Length: 7728394, dtype: int32

#### Timezone Column
> Removed `US/` prefix. e.g., `US/Eastern` → `Eastern`.


In [59]:
df2['Timezone']

0          US/Eastern
1          US/Eastern
2          US/Eastern
3          US/Eastern
4          US/Eastern
              ...    
7728389    US/Pacific
7728390    US/Pacific
7728391    US/Pacific
7728392    US/Pacific
7728393    US/Pacific
Name: Timezone, Length: 7728394, dtype: str

In [60]:
df2['Timezone'] = df2['Timezone'].str[3:]

In [61]:
df2['Timezone']

0          Eastern
1          Eastern
2          Eastern
3          Eastern
4          Eastern
            ...   
7728389    Pacific
7728390    Pacific
7728391    Pacific
7728392    Pacific
7728393    Pacific
Name: Timezone, Length: 7728394, dtype: str

#### Wind Direction Column
> Standardized case (`Calm` → `CALM`), then expanded abbreviations to full directional names.  
> e.g., `N` → `North`, `SW` → `Southwest`.


In [62]:
df2['Wind_Direction'].value_counts()

Wind_Direction
CALM        1136830
S            419989
SSW          384840
W            383913
WNW          378781
NW           369352
Calm         368557
SW           364470
WSW          353806
SSE          349110
NNW          333427
N            307151
SE           294901
E            278914
ESE          268311
NE           258639
ENE          258474
NNE          255311
VAR          250566
South        177375
West         164751
North        153385
Variable     113996
East         103545
Name: count, dtype: int64

In [63]:
df2['Wind_Direction'] = df2['Wind_Direction'].str.replace('Calm' , 'CALM')

In [64]:
df2['Wind_Direction'].value_counts()

Wind_Direction
CALM        1505387
S            419989
SSW          384840
W            383913
WNW          378781
NW           369352
SW           364470
WSW          353806
SSE          349110
NNW          333427
N            307151
SE           294901
E            278914
ESE          268311
NE           258639
ENE          258474
NNE          255311
VAR          250566
South        177375
West         164751
North        153385
Variable     113996
East         103545
Name: count, dtype: int64

In [65]:
df2['End_Time']

0          2016-02-08 11:00:00
1          2016-02-08 06:37:59
2          2016-02-08 07:19:27
3          2016-02-08 07:53:34
4          2016-02-08 08:09:07
                  ...         
7728389    2019-08-23 18:32:01
7728390    2019-08-23 19:38:23
7728391    2019-08-23 19:28:49
7728392    2019-08-23 19:29:42
7728393    2019-08-23 19:21:31
Name: End_Time, Length: 7728394, dtype: str

In [66]:
df2['Start_Time']

0          2016-02-08 05:46:00
1          2016-02-08 06:07:59
2          2016-02-08 06:49:27
3          2016-02-08 07:23:34
4          2016-02-08 07:39:07
                  ...         
7728389    2019-08-23 18:03:25
7728390    2019-08-23 19:11:30
7728391    2019-08-23 19:00:21
7728392    2019-08-23 19:00:21
7728393    2019-08-23 18:52:06
Name: Start_Time, Length: 7728394, dtype: str

In [67]:
df2['Accident_Time'] = df2['End_Time'].astype('datetime64[ns]') - df2['Start_Time'].astype('datetime64[ns]')

In [68]:
df2['Accident_Time']

0         0 days 05:14:00
1         0 days 00:30:00
2         0 days 00:30:00
3         0 days 00:30:00
4         0 days 00:30:00
                ...      
7728389   0 days 00:28:36
7728390   0 days 00:26:53
7728391   0 days 00:28:28
7728392   0 days 00:29:21
7728393   0 days 00:29:25
Name: Accident_Time, Length: 7728394, dtype: timedelta64[ns]

In [69]:
df2['End_Time'] = df2['End_Time'].astype('datetime64[ns]')
df2['Start_Time'] = df2['Start_Time'].astype('datetime64[ns]')

In [70]:
df2['End_Time']

0         2016-02-08 11:00:00
1         2016-02-08 06:37:59
2         2016-02-08 07:19:27
3         2016-02-08 07:53:34
4         2016-02-08 08:09:07
                  ...        
7728389   2019-08-23 18:32:01
7728390   2019-08-23 19:38:23
7728391   2019-08-23 19:28:49
7728392   2019-08-23 19:29:42
7728393   2019-08-23 19:21:31
Name: End_Time, Length: 7728394, dtype: datetime64[ns]

In [71]:
df['Start_Time']

0          2016-02-08 05:46:00
1          2016-02-08 06:07:59
2          2016-02-08 06:49:27
3          2016-02-08 07:23:34
4          2016-02-08 07:39:07
                  ...         
7728389    2019-08-23 18:03:25
7728390    2019-08-23 19:11:30
7728391    2019-08-23 19:00:21
7728392    2019-08-23 19:00:21
7728393    2019-08-23 18:52:06
Name: Start_Time, Length: 7728394, dtype: str

In [72]:
df.columns

Index(['ID', 'Source', 'Severity', 'Start_Time', 'End_Time', 'Start_Lat',
       'Start_Lng', 'End_Lat', 'End_Lng', 'Distance(mi)', 'Description',
       'Street', 'City', 'County', 'State', 'Zipcode', 'Country', 'Timezone',
       'Airport_Code', 'Weather_Timestamp', 'Temperature(F)', 'Wind_Chill(F)',
       'Humidity(%)', 'Pressure(in)', 'Visibility(mi)', 'Wind_Direction',
       'Wind_Speed(mph)', 'Precipitation(in)', 'Weather_Condition', 'Amenity',
       'Bump', 'Crossing', 'Give_Way', 'Junction', 'No_Exit', 'Railway',
       'Roundabout', 'Station', 'Stop', 'Traffic_Calming', 'Traffic_Signal',
       'Turning_Loop', 'Sunrise_Sunset', 'Civil_Twilight', 'Nautical_Twilight',
       'Astronomical_Twilight'],
      dtype='str')

In [73]:
## Handling Temprature

In [74]:
df2['Temperature(F)'] = round((df2['Temperature(F)']))

In [75]:
df2['Temperature(F)'].describe()

count    7.728394e+06
mean     6.171284e+01
std      1.881424e+01
min     -8.900000e+01
25%      5.000000e+01
50%      6.400000e+01
75%      7.600000e+01
max      2.070000e+02
Name: Temperature(F), dtype: float64

In [76]:
df2['Temperature(F)'] = df2['Temperature(F)'].astype('int32')

In [77]:
## Handling pressure

In [78]:
df2['Pressure(in)']


0          29.68
1          29.65
2          29.67
3          29.64
4          29.65
           ...  
7728389    28.92
7728390    29.39
7728391    29.74
7728392    29.62
7728393    28.63
Name: Pressure(in), Length: 7728394, dtype: float64

In [79]:
df2['Pressure(in)'] = round(df2['Pressure(in)'])

In [80]:
df2['Pressure(in)'] = df2['Pressure(in)'].astype('int16')

In [81]:
## Visisbility 

In [82]:
df2['Visibility(mi)'].describe()

count    7.728394e+06
mean     9.111221e+00
std      2.660819e+00
min      0.000000e+00
25%      1.000000e+01
50%      1.000000e+01
75%      1.000000e+01
max      1.400000e+02
Name: Visibility(mi), dtype: float64

In [83]:
df2['Visibility(mi)'] = round(df2['Visibility(mi)'])

In [84]:
df2['Visibility(mi)'] = df2['Visibility(mi)'].astype('int16')

In [85]:
## wind direction

In [86]:
df2['Wind_Direction'].value_counts()

Wind_Direction
CALM        1505387
S            419989
SSW          384840
W            383913
WNW          378781
NW           369352
SW           364470
WSW          353806
SSE          349110
NNW          333427
N            307151
SE           294901
E            278914
ESE          268311
NE           258639
ENE          258474
NNE          255311
VAR          250566
South        177375
West         164751
North        153385
Variable     113996
East         103545
Name: count, dtype: int64

In [87]:
mapping = {
    'S' : 'South',
    'W' : 'West',
    'E' : 'East',
    'N' : 'North',
    'SSW' : 'South South West',
    'WNW' : 'West North West',
    'NW' :'North West',
    'SW' :'South West',
    'WSW' : 'West South West',
    'SSE': 'South South East',
    'NNW' : 'North North West',
    'SE' : 'South East',
    'ESE' : 'East South East',
    'NE' : 'North East',
    'ENE' : 'East North East',
    'NNE' : 'North North East',
    'CALM' : 'CALM',
    'VAR' : 'Varaible',
    'Varaible' : 'Varaible',
    'East' : 'East',
    'West' : "West",
    'South' : 'South',
    'North' : 'North'
}

In [88]:
df2['Wind_Direction']

0          CALM
1          CALM
2            SW
3            SW
4            SW
           ... 
7728389       W
7728390      SW
7728391     SSW
7728392      SW
7728393      SW
Name: Wind_Direction, Length: 7728394, dtype: str

In [89]:
df2['Wind_Direction'] = df2['Wind_Direction'].map(mapping)

#### Start Time & End Time
> Converted to `datetime64` dtype. Calculated `Accident_Time` = `End_Time - Start_Time`.


In [65]:
df2['End_Time']

0          2016-02-08 11:00:00
1          2016-02-08 06:37:59
2          2016-02-08 07:19:27
3          2016-02-08 07:53:34
4          2016-02-08 08:09:07
                  ...         
7728389    2019-08-23 18:32:01
7728390    2019-08-23 19:38:23
7728391    2019-08-23 19:28:49
7728392    2019-08-23 19:29:42
7728393    2019-08-23 19:21:31
Name: End_Time, Length: 7728394, dtype: str

In [66]:
df2['Start_Time']

0          2016-02-08 05:46:00
1          2016-02-08 06:07:59
2          2016-02-08 06:49:27
3          2016-02-08 07:23:34
4          2016-02-08 07:39:07
                  ...         
7728389    2019-08-23 18:03:25
7728390    2019-08-23 19:11:30
7728391    2019-08-23 19:00:21
7728392    2019-08-23 19:00:21
7728393    2019-08-23 18:52:06
Name: Start_Time, Length: 7728394, dtype: str

In [67]:
df2['Accident_Time'] = df2['End_Time'].astype('datetime64[ns]') - df2['Start_Time'].astype('datetime64[ns]')

In [68]:
df2['Accident_Time']

0         0 days 05:14:00
1         0 days 00:30:00
2         0 days 00:30:00
3         0 days 00:30:00
4         0 days 00:30:00
                ...      
7728389   0 days 00:28:36
7728390   0 days 00:26:53
7728391   0 days 00:28:28
7728392   0 days 00:29:21
7728393   0 days 00:29:25
Name: Accident_Time, Length: 7728394, dtype: timedelta64[ns]

In [69]:
df2['End_Time'] = df2['End_Time'].astype('datetime64[ns]')
df2['Start_Time'] = df2['Start_Time'].astype('datetime64[ns]')

In [70]:
df2['End_Time']

0         2016-02-08 11:00:00
1         2016-02-08 06:37:59
2         2016-02-08 07:19:27
3         2016-02-08 07:53:34
4         2016-02-08 08:09:07
                  ...        
7728389   2019-08-23 18:32:01
7728390   2019-08-23 19:38:23
7728391   2019-08-23 19:28:49
7728392   2019-08-23 19:29:42
7728393   2019-08-23 19:21:31
Name: End_Time, Length: 7728394, dtype: datetime64[ns]

In [71]:
df['Start_Time']

0          2016-02-08 05:46:00
1          2016-02-08 06:07:59
2          2016-02-08 06:49:27
3          2016-02-08 07:23:34
4          2016-02-08 07:39:07
                  ...         
7728389    2019-08-23 18:03:25
7728390    2019-08-23 19:11:30
7728391    2019-08-23 19:00:21
7728392    2019-08-23 19:00:21
7728393    2019-08-23 18:52:06
Name: Start_Time, Length: 7728394, dtype: str

In [72]:
df.columns

Index(['ID', 'Source', 'Severity', 'Start_Time', 'End_Time', 'Start_Lat',
       'Start_Lng', 'End_Lat', 'End_Lng', 'Distance(mi)', 'Description',
       'Street', 'City', 'County', 'State', 'Zipcode', 'Country', 'Timezone',
       'Airport_Code', 'Weather_Timestamp', 'Temperature(F)', 'Wind_Chill(F)',
       'Humidity(%)', 'Pressure(in)', 'Visibility(mi)', 'Wind_Direction',
       'Wind_Speed(mph)', 'Precipitation(in)', 'Weather_Condition', 'Amenity',
       'Bump', 'Crossing', 'Give_Way', 'Junction', 'No_Exit', 'Railway',
       'Roundabout', 'Station', 'Stop', 'Traffic_Calming', 'Traffic_Signal',
       'Turning_Loop', 'Sunrise_Sunset', 'Civil_Twilight', 'Nautical_Twilight',
       'Astronomical_Twilight'],
      dtype='str')

---
### 5.3 Handling Accuracy & Validity


#### Temperature (F)
> Rounded to nearest integer and converted to `int32` for memory efficiency.


In [74]:
df2['Temperature(F)'] = round((df2['Temperature(F)']))

In [75]:
df2['Temperature(F)'].describe()

count    7.728394e+06
mean     6.171284e+01
std      1.881424e+01
min     -8.900000e+01
25%      5.000000e+01
50%      6.400000e+01
75%      7.600000e+01
max      2.070000e+02
Name: Temperature(F), dtype: float64

In [76]:
df2['Temperature(F)'] = df2['Temperature(F)'].astype('int32')

#### Pressure (in)
> Rounded to nearest integer and converted to `int16`.


In [78]:
df2['Pressure(in)']


0          29.68
1          29.65
2          29.67
3          29.64
4          29.65
           ...  
7728389    28.92
7728390    29.39
7728391    29.74
7728392    29.62
7728393    28.63
Name: Pressure(in), Length: 7728394, dtype: float64

In [79]:
df2['Pressure(in)'] = round(df2['Pressure(in)'])

In [80]:
df2['Pressure(in)'] = df2['Pressure(in)'].astype('int16')

#### Visibility (mi)
> Rounded to nearest integer and converted to `int16`.


In [81]:
## Visisbility 

In [82]:
df2['Visibility(mi)'].describe()

count    7.728394e+06
mean     9.111221e+00
std      2.660819e+00
min      0.000000e+00
25%      1.000000e+01
50%      1.000000e+01
75%      1.000000e+01
max      1.400000e+02
Name: Visibility(mi), dtype: float64

In [83]:
df2['Visibility(mi)'] = round(df2['Visibility(mi)'])

In [84]:
df2['Visibility(mi)'] = df2['Visibility(mi)'].astype('int16')

#### Wind Direction — Full Name Mapping
> All abbreviated direction codes mapped to full names for clarity.


In [85]:
## wind direction

In [86]:
df2['Wind_Direction'].value_counts()

Wind_Direction
CALM        1505387
S            419989
SSW          384840
W            383913
WNW          378781
NW           369352
SW           364470
WSW          353806
SSE          349110
NNW          333427
N            307151
SE           294901
E            278914
ESE          268311
NE           258639
ENE          258474
NNE          255311
VAR          250566
South        177375
West         164751
North        153385
Variable     113996
East         103545
Name: count, dtype: int64

In [87]:
mapping = {
    'S' : 'South',
    'W' : 'West',
    'E' : 'East',
    'N' : 'North',
    'SSW' : 'South South West',
    'WNW' : 'West North West',
    'NW' :'North West',
    'SW' :'South West',
    'WSW' : 'West South West',
    'SSE': 'South South East',
    'NNW' : 'North North West',
    'SE' : 'South East',
    'ESE' : 'East South East',
    'NE' : 'North East',
    'ENE' : 'East North East',
    'NNE' : 'North North East',
    'CALM' : 'CALM',
    'VAR' : 'Varaible',
    'Varaible' : 'Varaible',
    'East' : 'East',
    'West' : "West",
    'South' : 'South',
    'North' : 'North'
}

In [88]:
df2['Wind_Direction']

0          CALM
1          CALM
2            SW
3            SW
4            SW
           ... 
7728389       W
7728390      SW
7728391     SSW
7728392      SW
7728393      SW
Name: Wind_Direction, Length: 7728394, dtype: str

In [89]:
df2['Wind_Direction'] = df2['Wind_Direction'].map(mapping)

In [90]:
## Wind speed

In [91]:
df2['Wind_Speed(mph)'].describe()

count    7.728394e+06
mean     7.634823e+00
std      5.223725e+00
min      0.000000e+00
25%      5.000000e+00
50%      7.000000e+00
75%      1.000000e+01
max      1.087000e+03
Name: Wind_Speed(mph), dtype: float64

In [92]:
df2['Wind_Speed(mph)'] = df2['Wind_Speed(mph)'].clip(upper=300)

In [93]:
df2['Wind_Speed(mph)'].describe()

count    7.728394e+06
mean     7.633833e+00
std      5.119394e+00
min      0.000000e+00
25%      5.000000e+00
50%      7.000000e+00
75%      1.000000e+01
max      3.000000e+02
Name: Wind_Speed(mph), dtype: float64

In [98]:
df3 = df2.copy()

---
## ⚙️ 6. Feature Engineering

> Working on `df3` (copy of cleaned `df2`).  
> Goal: Create new meaningful columns that improve both analysis and ML model performance.

| Feature Group | New Columns | Purpose |
|---|---|---|
| **Time** | `hour`, `Day_Of_Week`, `Month_Name`, `Year`, `Season`, `is_weekend`, `time_of_day`, `is_Rush_Hour`, `Accident_Duration(min)` | Temporal patterns |
| **Severity** | `Severity` (label mapped) | Human-readable target |
| **Weather** | `Poor_Visibility`, `Is_Precipitating`, `Temp_Category` | Risk flags |
| **POI** | `Nearby_POI_Count`, `Has_Any_Poi` | Fix POI imbalance |

---


### 6.1 Time-Based Features

In [100]:
## Accident time
df3['Accident_Duration(min)'] = round((df3['End_Time'] - df3['Start_Time']).dt.total_seconds()/60)

In [101]:
## hour  
df3['hour'] = df3['Start_Time'].dt.hour

In [102]:
## day of week
df3['Day_Of_Week'] = df3['Start_Time'].dt.day_name()


In [103]:
## month
df3['Month_Name'] = df3['Start_Time'].dt.month_name()

In [104]:
## Year
df3['Year'] = df3['Start_Time'].dt.year

In [105]:
df3['Month_Number']=df3['Start_Time'].dt.month

In [106]:
## Season
season = {
    12:'Winter', 1:'Winter', 2:'Winter',
    3:'Spring',  4:'Spring', 5:'Spring',
    6:'Summer',  7:'Summer', 8:'Summer',
    9:'Fall',   10:'Fall',  11:'Fall'
}

In [107]:
df3['Season'] = df3['Month_Number'].map(season)

In [108]:
df3.drop(columns='Month_Number',inplace=True)

In [137]:
## is weekend
df3['is_weekend'] = df3['Start_Time'].dt.dayofweek >= 5

In [138]:
## Time of the day
df3['time_of_day'] =pd.cut(df3['hour'],
    bins=[0,6,12,17,21,24],
    labels=['Night','Morning','Afternoon','Evening','Late Night'])

In [139]:
## is rush hour
df3['is_Rush_Hour'] = df3['hour'].between(7,9) | df3['hour'].between(16,19)

### 6.2 Severity Label Mapping
> Mapped numeric severity (1–4) to descriptive labels: `Low`, `Medium`, `High`, `Critical`.


In [141]:
# ##is_sever
# df3['is_severe'] = df3['Severity'] >= 3

In [113]:
## severity
mapping = {
    1 : 'Low',
    2:'Medium',
    3 : 'High' ,
    4 : 'Critical'
}

In [114]:
## severity
df3['Severity'] = df3['Severity'].map(mapping)

### 6.3 Weather Risk Flags
> Binary flags to capture dangerous weather conditions clearly.


In [118]:
## Poor Visisbility
df3['Poor_Visibility'] = df3['Visibility(mi)'] < 1

In [120]:
## Is precipiting
df3['Is_Precipitaing'] = df3['Precipitation(in)'] > 0

In [121]:
## temprature
df3['Temp_Category'] = pd.cut(df['Temperature(F)'] , bins = [-50 , 32,60,85,150] , labels=['Freezing' , 'Cold' , 'Mild' , 'Hot'])

### 6.4 POI Aggregation
> All 12 POI boolean columns (heavily imbalanced individually) are combined into two meaningful features:
> - `Nearby_POI_Count` — total number of POIs near the accident
> - `Has_Any_Poi` — simple binary flag


In [124]:
## Nearby POI Count
cols = ['Amenity',
       'Bump', 'Crossing', 'Give_Way', 'Junction', 'No_Exit', 'Railway',
       'Roundabout', 'Station', 'Stop', 'Traffic_Calming', 'Traffic_Signal']

In [126]:
df3['Nearby_POI_Count'] = df[cols].sum(axis=1)

In [129]:
## has any poi
df3['Has_Any_Poi'] = df3['Nearby_POI_Count'] > 0

In [154]:
df3

,Severity,Start_Lat,Start_Lng,Distance(mi),City,State,Timezone,Temperature(F),Humidity(%),Pressure(in),...,Season,is_severe,Poor_Visibility,Is_Precipitaing,Temp_Category,Nearby_POI_Count,Has_Any_Poi,is_weekend,time_of_day,is_Rush_Hour
0,High,39.865147,-84.058723,0.010,Dayton,Ohio,Eastern,37,91.0,30,...,Winter,True,False,True,Cold,0,False,False,Night,False
1,Medium,39.928059,-82.831184,0.010,Reynoldsburg,Ohio,Eastern,38,100.0,30,...,Winter,False,False,False,Cold,0,False,False,Night,False
2,Medium,39.063148,-84.032608,0.010,Williamsburg,Ohio,Eastern,36,100.0,30,...,Winter,False,False,False,Cold,1,True,False,Night,False
3,High,39.747753,-84.205582,0.010,Dayton,Ohio,Eastern,35,96.0,30,...,Winter,True,False,False,Cold,0,False,False,Morning,True
4,Medium,39.627781,-84.188354,0.010,Dayton,Ohio,Eastern,36,89.0,30,...,Winter,False,False,False,Cold,1,True,False,Morning,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7728389,Medium,34.002480,-117.379360,0.543,Riverside,California,Pacific,86,40.0,29,...,Summer,False,False,False,Hot,0,False,False,Evening,True
7728390,Medium,32.766960,-117.148060,0.338,San Diego,California,Pacific,70,73.0,29,...,Summer,False,False,False,Mild,0,False,False,Evening,True
7728391,Medium,33.775450,-117.847790,0.561,Orange,California,Pacific,73,64.0,30,...,Summer,False,False,False,Mild,1,True,False,Evening,True
7728392,Medium,33.992460,-118.403020,0.772,Culver City,California,Pacific,71,81.0,30,...,Summer,False,False,False,Mild,0,False,False,Evening,True


---
## 💾 7. Saving Datasets

Two datasets saved:
- **`Cleaned_Dataset.csv`** — Full cleaned dataset with all original + new columns (for EDA & analysis)
- **`ML_Ready_Dataset.csv`** — Feature engineered version ready for ML modelling


In [156]:
df3.to_csv("Ml_Ready_Dataset.csv")

In [158]:
df2.to_csv("Clened_Dataset.csv")